### ISSUE TITLE
Support arrow keys and page up/down without modifier when in terminal navigation mode

### ISSUE THREAD

**Issue by jeanbrazcosta** [microsoft/vscode#149882]
When using command control + sdta upwards to read the terminal buffer, it is not possible to read the entire log.

Example when running a build of a java project with the mvn clean install command, thousands of log lines are produced when pressing control + up arrow I do not navigate through this entire report.

I think this command should only be pressed once and then it could be possible to navigate only using arrows and, advance or rewind a page when using pageup and pagedown and when pressing control plus down arrow it could exit the buffer area as it already does today.

I use the latest version of Nvda as well as the latest version of Vscode.

**Comment by jvesouza** [comment]
In my opinion, access to the terminal for those who use a screen reader in vscode is not very productive and comfortable, especially if there is a lot of information to be consulted.
I use orca as a screen reader and whenever I need to perform some operation in the terminal that generates a large output, I end up redirecting the command output to a file. Then I open this file in VScode to be able to analyze the result.
It seems to me that it would be interesting, as long as possible, to be able to open the terminal buffer in an editor.

**Comment by jvesouza** [comment]
1. Browse the terminal's contents by line, word and character using the arrows.
2. Select and copy parts of the content using the keyboard.
3. Activate link if there is one.
4. Run searches on the terminal's content.
5. Re-execute previously executed commands.

**Comment by jeanbrazcosta** [comment]
Another thing I can't do is complete a command in the terminal either.
For example, when typing git pu and pressing enter, I even hear NVDA say
push, but this is not completed in the command.

I tried to press spacebar after listening, press enter more then it
executes the wrong command because I haven't finished typing yet.

Em seg., 23 de mai. de 2022 às 14:27, José Vilmar Estácio de Souza <
***@***.***> escreveu:

**Comment by isidorn** [comment]
@jvesouza this is a great list
@Tyriar let me know if you would like me to help and create separate feature requests for some of these
@jeanbrazcosta I think there is already an issue in the terminal for this - suggestions not being read in the terminal.

**Comment by jvesouza** [comment]
> Thanks @Tyriar.
>

> @jvesouza <https://github.com/jvesouza>
>
>     Activate link if there is one.
>
> You can't do this while navigating the buffer but we do have commands 
> exposed to activate links now. Check out these commands in the command 
> palette:
>
>   * Terminal: Open Detected Link...
>
When using this command I was presented with a list with several texts 
but none of the texts presented was a link.


>   * Terminal: Open Last Url Link
>
When executing this command I didn't get any return, as if there were no 
links present in the terminal buffer.
>
>   * Terminal: Open Last Local File Link
>
This command also produced no response.


>     Re-execute previously executed commands.
>
> This is all up to the shell currently, how would you expect to be able 
> to do this?
>
> FWIW if you enable our experimental shell integration via 
> |"terminal.integrated.shellIntegration.enabled": true| you can use the 
> quick pick to search and run recent commands detected by it using a 
> keybinding like this:
>
> |{ "key": "ctrl+space", "command": 
> "workbench.action.terminal.runRecentCommand", "when": "terminalFocus" },|

> I can re-execute commands using the arrow keys to select the desired 
> command, as I normally do in an external terminal. The problem is that 
> I can't edit the selected command.
The command terminal.runRecentCommand works perfectly and if there was 
an option to edit the command before running it would be great.



>     Run searches on the terminal's content.
>
> We do have search in the terminal but I'm not actually sure how it 
> fares with a screen reader currently.
>
> ------------------------------------------------------------------------
>
> @jvesouza <https://github.com/jvesouza>
>
>     Another thing I can't do is complete a command in the terminal either.
>
> Normally this is done via tab which screen readers often override. Do 
> you know how other terminals handle this? One workaround you could try 
> is to setup a custom keybinding to send the tab sequence to the shell:
>
> {
> 	"key":"ctrl+.",
> 	"command":"workbench.action.terminal.sendSequence",
> 	"when":"terminalFocus",
> 	"args":{  "text":"\u0009"  }
> }
>
> Also we plan on investing in shell integration-based auto complete 
> sometime in the future which will give you access to a autocomplete 
> similar to the editor, as opposed to shell-based.
>
Pressing the tab key actually completes the command. The big problem for 
me is that I can't edit or review this command.

**Comment by Tyriar** [comment]
> When using this command I was presented with a list with several texts
but none of the texts presented was a link.

Open Detected Link and the other related commands will only search the viewport for links currently, perhaps that's the issue you're hitting?

**Comment by Tyriar** [comment]
> The command terminal.runRecentCommand works perfectly and if there was an option to edit the command before running it would be great.

Good to know, we actually debated whether this is needed. Created https://github.com/microsoft/vscode/issues/150375

> Pressing the tab key actually completes the command. The big problem for
me is that I can't edit or review this command.

Created https://github.com/microsoft/vscode/issues/150376

**Comment by Tyriar** [comment]
And here are some issues for the other problems:

- https://github.com/microsoft/vscode/issues/69795
- https://github.com/microsoft/vscode/issues/150379

**Comment by Tyriar** [comment]
I created https://github.com/microsoft/vscode/issues/151938 for some changes we hope to do in June, and this issue I renamed to capture what I think is the essence of the issue. Let me know if I missed anything

**Comment by jvesouza** [comment]
@meganrogge 
In my opinion this implementation was excellent because it is more comfortable to browse the terminal buffer. It would be great if we could also use the arrow keys to navigate by word and by character.

**Comment by meganrogge** [comment]
@jvesouza so glad to hear that. that one is on the plan as well :)

**Comment by jeanbrazcosta** [comment]
Thank you very much for the information.

I did the test here in the insiders version, but I still couldn't see all
the log that appears in the terminal.

I would say that pressing page up after using the control command plus up
arrow I can see about 300 lines and my output should have more than 2000
lines as it is the result of a build produced by the mvn clean install
command.

Below is the build information I tested.

VS Code version: Code - Insiders 1.69.0-insider
(e71b6105ebb4fbcfbf38a1b62bcbafdb048d3468, 2022-06-27T07:59:47.622Z)
OS version: Windows_NT x64 10.0.19044
Restricted Mode: No
Remote OS version: Linux x64 5.10.16.3-microsoft-standard-WSL2

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-4200U CPU @ 1.60GHz (4 x 2295)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization:
disabled_off<br>direct_rendering_display_compositor:
disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads:
enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw:
disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode:
enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl:
enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|7.88GB (2.68GB free)|
|Process Argv|--crash-reporter-id a45738e8-c9c4-445a-ad0d-5fa819e01fb6|
|Screen Reader|yes|
|VM|0%|

|Item|Value|
|---|---|
|Remote|WSL: ubuntu|
|OS|Linux x64 5.10.16.3-microsoft-standard-WSL2|
|CPUs|Intel(R) Core(TM) i5-4200U CPU @ 1.60GHz (4 x 2294)|
|Memory (System)|6.10GB (3.98GB free)|
|VM|0%|
</details><details><summary>Extensions (3)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-language-pack-pt-BR|MS-|1.69.6220922
remote-wsl|ms-|0.66.3
java-snippet|Wsc|0.0.6


</details><details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368:30146709
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythonvspyl392:30422396
pythontb:30258533
pythonvspyt551:30291412
pythonptprofiler:30281269
vshan820:30294714
pythondataviewer:30285072
vscod805cf:30301675
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
pythonvs932:30404738
wslgetstarted:30449409
vscscmwlcmt:30465136
cppdebug:30492333
pylanb8912cf:30496797
vsclangdf:30492506

```

</details>



Em seg., 27 de jun. de 2022 às 19:48, Daniel Imms ***@***.***>
escreveu:

**Comment by jvesouza** [comment]
@jeanbrazcosta 
There may be a limit on the number of lines stored in the terminal's buffer. When I run a command that has a very large output, I prefer to redirect that output to a file and open that file in VSCode.

**Comment by Tyriar** [comment]
There is a limit, it should be 1000 lines by default. It's set by the `terminal.integrated.scrollback` setting
